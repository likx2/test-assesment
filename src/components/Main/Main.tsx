import React, { FC } from 'react';
import {
    useUserPreferences
} from "../../context/UserPreferencesContext";

const Main: FC = () => {
    const {
        windowWidth,
        windowHeight,
        deviceType,
        theme,
        switchTheme
    } = useUserPreferences();

    return (
        <div>
            <p><strong>Window width:</strong> {windowWidth}</p>
            <p><strong>Window height:</strong> {windowHeight}</p>
            <p><strong>Device type:</strong> {deviceType}</p>
            <p><strong>Application theme:</strong> {theme}</p>
            <button onClick={() => switchTheme()}>Change theme</button>
        </div>
    );
};

export default Main;