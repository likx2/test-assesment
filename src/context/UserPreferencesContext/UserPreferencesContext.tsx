import React, { createContext, FC, ReactNode, useEffect, useState } from 'react';
import throttle from "lodash/throttle";
import { THEME } from "../../shared/types";

export interface UserPreferencesContextType {
    windowWidth: number,
    windowHeight: number,
    deviceType: string,
    theme: THEME,
    switchTheme: () => void;
}

export const UserPreferencesContext = createContext<UserPreferencesContextType | null>(null);

const THROTTLE_WAIT = 400;

interface UserPreferencesProviderProps {
    children: ReactNode;
}

export const UserPreferencesProvider: FC<UserPreferencesProviderProps> = ({ children }) => {
    const [windowWidth, setWindowWidth] = useState(window.innerWidth);
    const [windowHeight, setWindowHeight] = useState(window.innerHeight);
    const [theme, setTheme] = useState(THEME.LIGHT);

    useEffect(() => {
        window.addEventListener("resize", throttledResizeHandler);

        return () => {
            window.removeEventListener("resize", throttledResizeHandler);
        }
    }, []);

    const resizeHandler = () => {
        setWindowWidth(window.innerWidth);
        setWindowHeight(window.innerHeight);
    }

    const throttledResizeHandler = throttle(resizeHandler, THROTTLE_WAIT);

    const switchTheme = () => {
        if (theme === THEME.LIGHT) {
            setTheme(THEME.DARK);
        } else {
            setTheme(THEME.LIGHT);
        }
    }

    return (
        <UserPreferencesContext.Provider
            value={{
                windowWidth,
                windowHeight,
                deviceType: navigator.userAgent,
                theme,
                switchTheme,
            }}>
            {children}
        </UserPreferencesContext.Provider>
    );
};
