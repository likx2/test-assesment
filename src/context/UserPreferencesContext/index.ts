export * from "./UserPreferencesContext";
export { useUserPreferences } from "./useUserPreferences";