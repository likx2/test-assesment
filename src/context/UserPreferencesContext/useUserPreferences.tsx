import { useContext } from "react";
import { UserPreferencesContext, UserPreferencesContextType } from "./UserPreferencesContext";

export const useUserPreferences = () =>{
    return useContext(UserPreferencesContext) as UserPreferencesContextType;
}