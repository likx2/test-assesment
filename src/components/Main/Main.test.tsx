import React from "react";
import { screen, render, fireEvent } from "@testing-library/react";
import Main from "./Main";
import { UserPreferencesProvider } from "../../context/UserPreferencesContext";
import { THEME } from "../../shared/types";

describe("<Main />", () => {
    it("should match the snapshot", () => {
        const { asFragment } = render(
            <UserPreferencesProvider>
                <Main/>
            </UserPreferencesProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render user preferences", () => {
        const testDeviceType = "Macbook Pro M2";
        const previousDeviceType = window.navigator.userAgent;
        const testWindowWidth = 600;
        const previousInnerWidth = window.innerWidth;
        const testWindowHeight = 500;
        const previousInnerHeight = window.innerHeight;

        Object.defineProperty(window.navigator, "userAgent", {
            configurable: true,
            value: testDeviceType,
        });

        Object.defineProperty(window, "innerWidth", {
            configurable: true,
            value: testWindowWidth,
        });

        Object.defineProperty(window, "innerHeight", {
            configurable: true,
            value: testWindowHeight,
        });

        render(
            <UserPreferencesProvider>
                <Main/>
            </UserPreferencesProvider>
        );

        expect(screen.getByText(testDeviceType)).toBeInTheDocument();
        expect(screen.getByText(testWindowWidth)).toBeInTheDocument();
        expect(screen.getByText(testWindowHeight)).toBeInTheDocument();
        expect(screen.getByText(THEME.LIGHT)).toBeInTheDocument();

        Object.defineProperty(window.navigator, "userAgent", {
            configurable: true,
            value: previousDeviceType,
        });

        Object.defineProperty(window, "innerWidth", {
            configurable: true,
            value: previousInnerWidth,
        });

        Object.defineProperty(window, "innerHeight", {
            configurable: true,
            value: previousInnerHeight,
        });
    });

    it("should update theme", () => {
        render(
            <UserPreferencesProvider>
                <Main/>
            </UserPreferencesProvider>
        );

        expect(screen.getByText(THEME.LIGHT)).toBeInTheDocument();

        const button = screen.getByRole("button");

        fireEvent.click(button);

        expect(screen.getByText(THEME.DARK)).toBeInTheDocument();
    });
});