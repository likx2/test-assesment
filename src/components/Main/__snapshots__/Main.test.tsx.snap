// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <p>
      <strong>
        Window width:
      </strong>
       1024
    </p>
    <p>
      <strong>
        Window height:
      </strong>
       768
    </p>
    <p>
      <strong>
        Device type:
      </strong>
       Mozilla/5.0 (darwin) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0
    </p>
    <p>
      <strong>
        Application theme:
      </strong>
       light
    </p>
    <button>
      Change theme
    </button>
  </div>
</DocumentFragment>
`;
